### PR TITLE
Az602 - Update ELevelDB Backend to Support 2i (and AZ639)

### PR DESCRIPTION
**Pull request by rustyio**
## Overview
- This change adds 2i support to ELevelDB backend. 
- This required changes to the DELETE operation in `riak_kv_vnode`, which cascaded to changes in other backends. Specifically, the vnode needed to supply index specs to Mod:delete/N.
- Finally, this also includes AZ639, which makes the $bucket and $key fields implicit, rather than automatically generating them and tacking them on to the object metadata. Needed to do this in order to fully test the ELevelDB changes.
## Setup

Make sure that you have set the `storage_backend` config setting to `riak_kv_eleveldb_backend`.
### Index some documents...

``` bash
curl -v -X PUT \
-d 'data1' \
-H "x-riak-index-field1_bin: val1" \
-H "x-riak-index-field2_int: 1001" \
http://127.0.0.1:8098/riak/mybucket/mykey1

curl -v -X PUT \
-d 'data2' \
-H "x-riak-index-Field1_bin: val2" \
-H "x-riak-index-Field2_int: 1002" \
http://127.0.0.1:8098/riak/mybucket/mykey2

curl -v -X PUT \
-d 'data3' \
-H "X-RIAK-INDEX-FIELD1_BIN: val3" \
-H "X-RIAK-INDEX-FIELD2_INT: 1003" \
http://127.0.0.1:8098/riak/mybucket/mykey3

curl -v -X PUT \
-d 'data4' \
-H "x-riak-index-field1_bin: val4" \
-H "x-riak-index-field2_int: 1004" \
http://127.0.0.1:8098/riak/mybucket/mykey4
```
### Retrieve the documents...

``` bash
curl -i http://127.0.0.1:8098/riak/mybucket/mykey1
curl -i http://127.0.0.1:8098/riak/mybucket/mykey2
curl -i http://127.0.0.1:8098/riak/mybucket/mykey3
curl -i http://127.0.0.1:8098/riak/mybucket/mykey4
```
### Query against a binary field...

``` bash
curl http://localhost:8098/buckets/mybucket/index/field1_bin/val1
curl http://localhost:8098/buckets/mybucket/index/field1_bin/val2/val4
```
### Query against an integer field...

``` bash
curl http://localhost:8098/buckets/mybucket/index/field2_int/1001
curl http://localhost:8098/buckets/mybucket/index/field2_int/1002/1004
```
### Try some special queries...

``` bash
curl http://localhost:8098/buckets/mybucket/index/\$bucket/mybucket
curl http://localhost:8098/buckets/mybucket/index/\$key/mykey1/mykey3
```
### Should throw errors: too few / too many args...

``` bash
curl http://localhost:8098/buckets/mybucket/index/field2_int
curl http://localhost:8098/buckets/mybucket/index/field2_int/1/2/3
```
### Delete an object, make sure we can't find it...

```
curl -i -X DELETE http://127.0.0.1:8098/riak/mybucket/mykey1
curl -i http://127.0.0.1:8098/riak/mybucket/mykey1
curl http://localhost:8098/buckets/mybucket/index/field1_bin/val1
curl http://localhost:8098/buckets/mybucket/index/field2_int/1001
```

### CODE REVIEW

**Comment by beerriot** [comment]
Is there also a PR to add the `IndexSpecs` parameter to `riak_kv_innostore_backend:delete`?

File: src/riak_kv_eleveldb_backend.erl [hunk 1]
```diff
@@ -234,61 +290,90 @@ config_value(Key, Config) ->
...
     end.
 
+
 %% @private
 %% Return a function to fold over keys on this backend
 fold_keys_fun(FoldKeysFun, undefined) ->
-    fun(BK, Acc) ->
-            {Bucket, Key} = sext:decode(BK),
-            FoldKeysFun(Bucket, Key, Acc)
+    %% Fold across a specific bucket...
```
Comment (beerriot) [inline src/riak_kv_eleveldb_backend.erl]: This looks like a copy-paste-o: this is the "fold across everything" case, yes?

**Comment by beerriot** [comment]
Looks like `riak_kv_index_backend` needs the same modification (compile fails with behavior warning).

File: src/riak_kv_eleveldb_backend.erl (hunk 1 above)
Comment (rustyio) [inline src/riak_kv_eleveldb_backend.erl]: Fixed in a new commit to this pull request.

**Comment by rustyio** [comment]
Fix `riak_kv_index_backend` so that it supports delete/4. Keep in mind that we've already merged a pull request to delete riak_kv_index_backend.

**Comment by rustyio** [comment]
Waiting until this pull request lands: https://github.com/basho/innostore/pull/6 to update Innostore backend for new delete/4 function

Task is captured as AZ647.

**Comment by beerriot** [comment]
+1 merge, cool stuff!

**Pull request merged**